### PR TITLE
Add an ability to create a workspace on the server

### PR DESCRIPTION
**Pull request by MoriArchi**
The first need to create a workspace appears when a user registered in our app.

Implement the flow to create a workspace:

Create api endpoint: PUT /workspaces
Take userID from the token. (When user authorization will be
implemented)

Create workspace service, entity and repository

Fix typings for Express request

### CODE REVIEW

File: backend/tsconfig.json [hunk 1]
```diff
@@ -23,9 +23,9 @@
     "emitDecoratorMetadata": true,
     "resolveJsonModule": true,
     "esModuleInterop": true,
-    "types": ["node"],
+    "types": ["node", "./src/common/@types/express"],
```
Comment (MoriArchi) [inline backend/tsconfig.json:26]: It fix this error: http://prntscr.com/ttxhl6

File: backend/src/api/routes/index.ts [hunk 2]
```diff
@@ -1,8 +1,10 @@
 import { Express } from 'express';
 import userRoute from './userRoute';
+import workspaceRoute from './workspaceRoute';
 
 const routes = (app: Express) => {
   app.use('/api/users', userRoute);
+  app.use('/api', workspaceRoute);
```
Comment (AlikRakhmonov) [inline backend/src/api/routes/index.ts:7]:
```suggestion
  app.use('/api/workspaces', workspaceRoute);
```

there is an example above

File: backend/src/api/routes/workspaceRoute.ts [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import { Router, Request } from 'express';
+import { run } from '../../common/utils/routeHelper';
+import { createWorkspace } from '../../services/workspaceService';
+
+const router = Router();
+
+router
+  .put('/workspaces', run((req: Request) => createWorkspace({ name: req.body.name, userId: req.user.id })));
```
Comment (AlikRakhmonov) [inline backend/src/api/routes/workspaceRoute.ts:8]:
```suggestion
  .post('/', run((req: Request) => createWorkspace({ name: req.body.name, userId: req.user.id })));
```

File: backend/src/common/@types/express/index.d.ts [hunk 4]
```diff
@@ -3,10 +3,8 @@ import { IAuthUser } from "../../models/user";
...
 declare global {
   namespace Express {
-    interface User extends IAuthUser {}
-
     interface Request {
-      user?: User;
+      user?: IAuthUser;
       id?: string;
     }
   }
```
Comment (AlikRakhmonov) [inline backend/src/common/@types/express/index.d.ts:10]: do not change this file. rollback it

File: backend/src/common/models/workspace.ts [hunk 5]
```diff
@@ -0,0 +1,8 @@
+export interface ICreateWorkspace {
+  name: string;
+  userId: string;
+}
+
+export interface IWorkspaceResponse extends ICreateWorkspace {
+  id: string;
+}
```
Comment (AlikRakhmonov) [inline backend/src/common/models/workspace.ts:8]: one interface = one file
create a folder "workspace"

File: backend/src/data/entities/Workspace.ts [hunk 6]
```diff
@@ -0,0 +1,11 @@
+import { Entity, Column } from 'typeorm';
+import { AbstractEntity } from '../abstract/AbstractEntity';
+
+@Entity()
+export class Workspace extends AbstractEntity {
+  @Column()
+  name: string;
+
+  @Column()
+  userId: string;
```
Comment (AlikRakhmonov) [inline backend/src/data/entities/Workspace.ts:10]: add many-to-many relationship between workspace and user

File: backend/src/data/entities/Workspace.ts [hunk 7]
```diff
@@ -0,0 +1,11 @@
+import { Entity, Column } from 'typeorm';
+import { AbstractEntity } from '../abstract/AbstractEntity';
+
+@Entity()
+export class Workspace extends AbstractEntity {
+  @Column()
+  name: string;
```
Comment (AlikRakhmonov) [inline backend/src/data/entities/Workspace.ts:7]: make in unique and max length 100

File: backend/src/data/repositories/workspaceRepository.ts [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+import { Workspace } from '../entities/Workspace';
+import { ICreateWorkspace, IWorkspaceResponse } from '../../common/models/workspace';
+
+@EntityRepository(Workspace)
+export class WorkspaceRepository extends Repository<Workspace> {
+  async addWorkspace(data: ICreateWorkspace): Promise<IWorkspaceResponse> {
+    const workspace = this.create(data);
+
+    await workspace.save();
+    return workspace;
```
Comment (AlikRakhmonov) [inline backend/src/data/repositories/workspaceRepository.ts:11]: workspace has Workspace type, not IWorkspaceResponse. You should map it in service

File: backend/src/data/repositories/userRepository.ts [hunk 9]
```diff
@@ -1,15 +1,14 @@
 import { EntityRepository, Repository } from 'typeorm';
 
-import { User } from '../entities/UserEntity';
+import { User } from '../entities/User';
 import { ICreateUser } from '../../common/models/user/ICreateUser';
 
 @EntityRepository(User)
 class UserRepository extends Repository<User> {
-  async addUser(data: ICreateUser): Promise<User> {
```
Comment (MoriArchi) [inline backend/src/data/repositories/userRepository.ts:8]: We already use async at service

**Comment by MoriArchi** [comment]
Done

File: backend/src/api/routes/index.ts (hunk 2 above)
Comment (MoriArchi) [inline backend/src/api/routes/index.ts]: Сhanges: https://github.com/BinaryStudioAcademy/bsa-2020-chatito/pull/14/files

File: backend/src/common/models/workspace/workspaceResponse.ts [hunk 10]
```diff
@@ -0,0 +1,5 @@
+export interface IWorkspaceResponse {
+  workspaceId: string;
+  name: string;
+  createdByUserId: string;
+}
```
Comment (AlikRakhmonov) [inline backend/src/common/models/workspace/workspaceResponse.ts:5]: file name should be the same as interface 'I...'
do the same for all

Comment (MoriArchi) [reply backend/src/common/models/workspace/workspaceResponse.ts]: https://github.com/BinaryStudioAcademy/bsa-2020-chatito/pull/14/commits/d60a0ac7a61cf41a25aa888936b806125a7cf3ef

**Pull request merged**